File: slp/backend-dotnet/Features/Question/QuestionController.cs
using backend_dotnet.Features.Auth;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System;
using System.Security.Claims;
using System.Threading.Tasks;

namespace backend_dotnet.Features.Question;

[Authorize]
[ApiController]
[Route("api/[controller]")]
public class QuestionController : ControllerBase
{
    private readonly IQuestionService _questionService;
    private readonly IAuthService _authService;

    public QuestionController(IQuestionService questionService, IAuthService authService)
    {
        _questionService = questionService;
        _authService = authService;
    }

    private int? CurrentUserId => User.Identity?.IsAuthenticated == true
        ? int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)?.Value ?? "0")
        : null;

    private bool IsAdmin => User.IsInRole("admin");

    [HttpGet("{id}")]
    public async Task<IActionResult> GetQuestion(int id)
    {
        var question = await _questionService.GetQuestionByIdAsync(id);
        if (question == null)
            return NotFound();
        return Ok(question);
    }

    [HttpGet]
    public async Task<IActionResult> GetQuestions(
    [FromQuery] string? search,
    [FromQuery] string? type,
    [FromQuery] List<string>? tags,
    [FromQuery] bool? mine,
    [FromQuery] int page = 1,
    [FromQuery] int pageSize = 20)
    {
        // Enforce max page size
        if (page < 1) page = 1;
        if (pageSize < 1) pageSize = 20;
        if (pageSize > 50) pageSize = 50;

        var searchDto = new QuestionSearchDto
        {
            SearchTerm = search,
            Type = type,
            Tags = tags,
            UserId = mine == true ? CurrentUserId : null
        };

        var results = await _questionService.SearchQuestionsAsync(searchDto, page, pageSize);
        return Ok(results);
    }

    [HttpPost]
    public async Task<IActionResult> CreateQuestion([FromBody] CreateQuestionDto dto)
    {
        if (!CurrentUserId.HasValue)
            return Unauthorized();

        var question = await _questionService.CreateQuestionAsync(CurrentUserId.Value, dto);
        return CreatedAtAction(nameof(GetQuestion), new { id = question.Id }, question);
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateQuestion(int id, [FromBody] UpdateQuestionDto dto)
    {
        if (!CurrentUserId.HasValue)
            return Unauthorized();

        var updated = await _questionService.UpdateQuestionAsync(id, CurrentUserId.Value, dto);
        if (updated == null)
            return NotFound();
        return Ok(updated);
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteQuestion(int id)
    {
        if (!CurrentUserId.HasValue)
            return Unauthorized();

        var deleted = await _questionService.DeleteQuestionAsync(id, CurrentUserId.Value, IsAdmin);
        if (!deleted)
            return NotFound();
        return NoContent();
    }
}

File: slp/backend-nestjs/src/modules/question/question.controller.ts
// src/modules/question/question.controller.ts

import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  UseGuards,
  HttpCode,
  HttpStatus,
  UnauthorizedException,
  NotFoundException,
  ForbiddenException,
} from "@nestjs/common";
import {
  ApiTags,
  ApiBearerAuth,
  ApiQuery,
  ApiResponse,
  ApiOperation,
} from "@nestjs/swagger";
import { QuestionService } from "./question.service";
import { CreateQuestionDto } from "./dto/create-question.dto";
import { UpdateQuestionDto } from "./dto/update-question.dto";
import { QuestionDto, QuestionListDto } from "./dto/question.dto";
import { QuestionSearchDto } from "./dto/question-search.dto";
import { PaginatedQuestionListDto } from "./dto/paginated-question-list.dto";
import { SessionGuard } from "../../common/guards/session.guard";
import { User } from "../../common/decorators/user.decorator";
import { isAdmin } from "../../helpers/admin.helper";

@ApiTags("questions")
@ApiBearerAuth()
@Controller("api/question")
@UseGuards(SessionGuard)
export class QuestionController {
  constructor(private readonly questionService: QuestionService) {}

  @Get(":id")
  @ApiOperation({ summary: "Get question by ID" })
  @ApiResponse({ status: 200, type: QuestionDto })
  @ApiResponse({ status: 404, description: "Question not found" })
  async getQuestion(@Param("id") id: string, @User() user: any) {
    const question = await this.questionService.getQuestionById(+id);
    if (!question) throw new NotFoundException("Question not found");
    return question;
  }

  @Get()
  @ApiOperation({ summary: "Get questions with search/filter" })
  @ApiQuery({ name: "search", required: false })
  @ApiQuery({ name: "type", required: false })
  @ApiQuery({ name: "tags", required: false, isArray: true, type: String })
  @ApiQuery({ name: "mine", required: false, type: Boolean })
  @ApiQuery({ name: "page", required: false, type: Number, example: 1 })
  @ApiQuery({ name: "pageSize", required: false, type: Number, example: 20 })
  @ApiResponse({ status: 200, type: PaginatedQuestionListDto })
  async getQuestions(
    @Query("search") search?: string,
    @Query("type") type?: string,
    @Query("tags") tags?: string | string[],
    @Query("mine") mine?: boolean,
    @Query("page") page = 1,
    @Query("pageSize") pageSize = 20,
    @User() user?: any,
  ) {
    // Parse tags from query (could be string or array)
    let tagArray: string[] | undefined;
    if (tags) {
      tagArray = Array.isArray(tags) ? tags : [tags];
    }

    const searchDto: QuestionSearchDto = {
      searchTerm: search,
      type: type,
      tags: tagArray,
      userId: mine === true ? user?.id : undefined,
    };

    // Apply pagination constraints
    const effectivePage = Math.max(1, page);
    const effectivePageSize = Math.min(50, Math.max(1, pageSize));

    const result = await this.questionService.searchQuestions(
      searchDto,
      effectivePage,
      effectivePageSize,
    );
    // result is PaginatedResult<QuestionListDto> which matches shape of PaginatedQuestionListDto
    return new PaginatedQuestionListDto(
      result.items,
      result.total,
      result.page,
      result.pageSize,
    );
  }

  @Post()
  @ApiOperation({ summary: "Create a new question" })
  @ApiResponse({ status: 201, type: QuestionDto })
  async createQuestion(@Body() dto: CreateQuestionDto, @User() user: any) {
    if (!user) throw new UnauthorizedException();
    const question = await this.questionService.createQuestion(user.id, dto);
    return question;
  }

  @Put(":id")
  @ApiOperation({ summary: "Update a question" })
  @ApiResponse({ status: 200, type: QuestionDto })
  @ApiResponse({ status: 403, description: "Not owner" })
  @ApiResponse({ status: 404, description: "Question not found" })
  async updateQuestion(
    @Param("id") id: string,
    @Body() dto: UpdateQuestionDto,
    @User() user: any,
  ) {
    if (!user) throw new UnauthorizedException();
    const updated = await this.questionService.updateQuestion(
      +id,
      user.id,
      dto,
    );
    if (!updated) throw new NotFoundException("Question not found");
    return updated;
  }

  @Delete(":id")
  @ApiOperation({ summary: "Delete a question (admin or owner)" })
  @ApiResponse({ status: 204, description: "Deleted" })
  @ApiResponse({ status: 403, description: "Not authorized" })
  @ApiResponse({ status: 404, description: "Question not found" })
  @HttpCode(HttpStatus.NO_CONTENT)
  async deleteQuestion(@Param("id") id: string, @User() user: any) {
    if (!user) throw new UnauthorizedException();
    const isAdminUser = isAdmin(user.id); // Hardcoded admin check
    const deleted = await this.questionService.deleteQuestion(
      +id,
      user.id,
      isAdminUser,
    );
    if (!deleted) throw new NotFoundException("Question not found");
  }
}
